### PR TITLE
[core] feat(TagInput): resizable input that scales as you type

### PR DESCRIPTION
**Pull request by JosephChotard**
#### Checklist

- [N/A] Includes tests
- [N/A] Update documentation

#### Changes proposed in this pull request:

Currently the `TagInput` component (used by `MultiSelect2`) does not expand to accomodate the text the user is typing. This is because `input` elements don't scale with their input. This causes annoying situations where the user cannot see what they are typing when there is space to display it. 
To address this issue, I created a small `ResizeableInput` component (could be included in `TagInput` but this makes the intent clearer). This new component acts like a regular `input` element for every other component but it has some coding magic that causes it to fit the size of it's content. It will never grow bigger than it's closest parent with a `position: relative`. 

#### Reviewers should focus on:

This applies to both `TagInput` and `MultiSelect`
I checked that:
* The input never stretches wider than the parent
* The input scales with it's content
* The input is never smaller than 80px
* The existing props still work (large, disabled, add on blur, ...)

This PR does not introduce changes to the API and is fully backwards compatible.

#### Screenshot

Old Input scaling:
numbers - old
New Input scaling:
numbers - new

Old input text being cut off
wrap old

New input text wrapping
wrap new

### CODE REVIEW

File: packages/core/src/components/tag-input/resizeableInput.tsx [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+ */
+
+import React, { forwardRef, useEffect, useRef, useState } from "react";
+
+import { HTMLInputProps } from "../../common";
+
+type ResizeableInputProps = HTMLInputProps;
+export type Ref = HTMLInputElement;
+
+export const ResizeableInput = forwardRef<Ref, ResizeableInputProps>(function ResizeableInput(props, ref) {
```
Comment (adidahiya) [inline packages/core/src/components/tag-input/resizeableInput.tsx]: I believe "resizable" is more correct, at least for US English. We use that elsewhere in Blueprint, so please rename all references to "resizeable" -> "resizable" (including the file name).

File: packages/core/src/components/tag-input/resizeableInput.tsx [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    const [content, setContent] = useState("");
+    const [width, setWidth] = useState(0);
+    const span = useRef<HTMLSpanElement>(null);
+
+    useEffect(() => {
+        if (span.current) {
```
Comment (adidahiya) [inline packages/core/src/components/tag-input/resizeableInput.tsx]: nit: please use explicit boolean coercion

```suggestion
        if (span.current != null) {
```

File: packages/core/src/components/tag-input/resizeableInput.tsx [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+            setWidth(span.current.offsetWidth);
+        }
+    }, [content]);
+
+    const { onChange, ...otherProps } = props;
+
+    const changeHandler: React.ChangeEventHandler<HTMLInputElement> = evt => {
```
Comment (adidahiya) [inline packages/core/src/components/tag-input/resizeableInput.tsx]: nit: use `handle...` naming for event callbacks

```suggestion
    const handleInputChange: React.ChangeEventHandler<HTMLInputElement> = evt => {
```

File: packages/core/src/components/tag-input/resizeableInput.tsx [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        setContent(evt.target.value);
+        onChange?.(evt);
+    };
+
+    return (
+        <span>
+            <span
+                id="hide"
+                ref={span}
+                style={{
```
Comment (adidahiya) [inline packages/core/src/components/tag-input/resizeableInput.tsx]: prefer styling with CSS rather than inline styles. you should create a `_resizableInput.scss` file next to this one and create a new class following our existing naming conventions...

```scss
.#{$ns}-resizable-input-span {
  max-height: 0;
  max-width: 100%;
  min-width: 80px;
  position: absolute;
  white-space: nowrap;
  z-index: -100;
}
```

File: packages/core/src/components/tag-input/resizeableInput.tsx [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+                    maxHeight: "0",
+                    maxWidth: "100%",
+                    minWidth: "80px",
+                    position: "absolute",
+                    whiteSpace: "nowrap",
+                    zIndex: -100,
```
Comment (adidahiya) [inline packages/core/src/components/tag-input/resizeableInput.tsx]: why is this necessary? can it be simply `-1`? can you leave a code comment?

File: packages/core/src/components/tag-input/resizeableInput.tsx [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+                }}
+            >
+                {/* Need to replace spaces with the html character for them to be preserved */}
```
Comment (adidahiya) [inline packages/core/src/components/tag-input/resizeableInput.tsx]: :+1:

**Review by adidahiya — CHANGES_REQUESTED** [review]
Nice, this seems to work really well in the [docs preview build](https://output.circle-artifacts.com/output/job/5d65e360-4486-435a-998d-09b9bccb2b61/artifacts/0/packages/docs-app/dist/index.html#core/components/tag-input)!

Looks like the tests are failing because we're using shallow rendering in many of the TagInput tests, and the `<ResizeableInput>` children don't get rendered. We can fix that by switching to `mount()` instead of `shallow()`.

**Comment by JosephChotard** [comment]
https://github.com/enzymejs/enzyme/issues/1473
https://github.com/enzymejs/enzyme/issues/1081
Changed the tests to use `mount` and modified a few `.simulate` calls as they caused the tests to hang (not sure why), instead I call the prop directly, I don't believe we need to test the event passing as that's React's responsibility.

File: packages/core/src/components/tag-input/resizeableInput.tsx (hunk 1 above)
Comment (JosephChotard) [reply packages/core/src/components/tag-input/resizeableInput.tsx]: ack

File: packages/core/src/components/tag-input/resizeableInput.tsx (hunk 5 above)
Comment (JosephChotard) [reply packages/core/src/components/tag-input/resizeableInput.tsx]: whoops, fixing it

**Comment by JosephChotard** [comment]
@adidahiya done all the changes, lmk if there's anything else

File: packages/core/src/common/classes.ts [hunk 7]
```diff
@@ -163,6 +163,8 @@ export const INPUT_GROUP = `${INPUT}-group`;
 export const INPUT_LEFT_CONTAINER = `${INPUT}-left-container`;
 export const INPUT_ACTION = `${INPUT}-action`;
 
+export const INPUT_RESIZABLE_INPUT_SPAN = `${NS}-resizable-input-span`;
```
Comment (adidahiya) [inline packages/core/src/common/classes.ts]: nit: the variable name should match the class name more closely:

```suggestion
export const RESIZABLE_INPUT_SPAN = `${NS}-resizable-input-span`;
```

File: packages/core/src/components/_index.scss [hunk 8]
```diff
@@ -37,6 +37,7 @@
 @import "tabs/tabs";
 @import "tag/tag";
 @import "tag-input/tag-input";
+@import "tag-input/resizeable-input";
```
Comment (adidahiya) [inline packages/core/src/components/_index.scss]: missed a file name here:

```suggestion
@import "tag-input/resizable-input";
```

File: packages/core/src/components/tag-input/resizableInput.tsx [hunk 9]
```diff
@@ -0,0 +1,39 @@
+/* !
+ * (c) Copyright 2022 Palantir Technologies Inc. All rights reserved.
+ */
+
+import React, { forwardRef, useEffect, useRef, useState } from "react";
+
+import { Classes, HTMLInputProps } from "../../common";
+
+type ResizableInputProps = HTMLInputProps;
+export type Ref = HTMLInputElement;
```
Comment (adidahiya) [inline packages/core/src/components/tag-input/resizableInput.tsx:9]: I don't think this typedef needs to exist. Just inline the type: `forwardRef<HTMLInputElement, ResizableInputProps>`

File: packages/core/src/components/tag-input/resizableInput.tsx [hunk 10]
```diff
@@ -0,0 +1,39 @@
+/* !
+ * (c) Copyright 2022 Palantir Technologies Inc. All rights reserved.
+ */
+
+import React, { forwardRef, useEffect, useRef, useState } from "react";
+
+import { Classes, HTMLInputProps } from "../../common";
+
+type ResizableInputProps = HTMLInputProps;
```
Comment (adidahiya) [inline packages/core/src/components/tag-input/resizableInput.tsx]: nit: export for consistency

```suggestion
export type ResizableInputProps = HTMLInputProps;
```

File: packages/core/test/tag-input/tagInputTests.tsx [hunk 11]
```diff
@@ -32,19 +26,17 @@ import { Button, Classes, Intent, Keys, Tag, TagInput, TagInputProps } from "../
...
 describe("<TagInput>", () => {
     it("passes inputProps to input element", () => {
         const onBlur = sinon.spy();
-        const input = shallow(<TagInput values={VALUES} inputProps={{ autoFocus: true, onBlur }} />).find("input");
+        const input = mount(<TagInput values={VALUES} inputProps={{ autoFocus: true, onBlur }} />).find("input");
         assert.isTrue(input.prop("autoFocus"));
         // check that event handler is proxied
         const fakeEvent = { flag: "yes" };
-        input.simulate("blur", fakeEvent);
+        input.prop("onBlur")?.(fakeEvent as any);
```
Comment (adidahiya) [inline packages/core/test/tag-input/tagInputTests.tsx:39]: I don't love this change but it appears to be the best path forward given the Enzyme issues you just linked. I'm surprised to learn that `.simulate()` is no longer recommended. And at some point I will move away from Enzyme anyway so these tests will get updated...

**Review by adidahiya — CHANGES_REQUESTED** [review]
looking good, just a few more minor things

File: packages/core/test/tag-input/tagInputTests.tsx (hunk 11 above)
Comment (JosephChotard) [reply packages/core/test/tag-input/tagInputTests.tsx:39]: agreed

**Comment by JosephChotard** [comment]
@adidahiya thanks for all the feedback, should be addressed now

**Review by adidahiya — APPROVED** [review]
LGTM, thanks for the thorough PR @JosephChotard :+1:

**Comment by bvandercar-vt** [comment]
@adidahiya this change has created a bug where the `placeholder` is no longer present for a `MultiSelect2`

**Comment by nerdstep** [comment]
Looks like there might be a regression with the input width not expanding to the full width of the component when using the `fill` prop:

image

File: packages/core/src/components/tag-input/resizableInput.tsx [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+    const { onChange, ...otherProps } = props;
+
+    const handleInputChange: React.ChangeEventHandler<HTMLInputElement> = evt => {
+        onChange?.(evt);
+        setContent(evt?.target?.value ?? "");
+    };
+
+    return (
+        <span>
+            <span id="hide" ref={span} className={Classes.RESIZABLE_INPUT_SPAN}>
```
Comment (bvandrc) [inline packages/core/src/components/tag-input/resizableInput.tsx:31]: Not sure why this `id` prop is needed / where it's used, but it's causing aXe a11y failures because all `id`s must be **unique**.

Must either remove `id` prop, or use `uniqueId` for it.

Comment (bvandrc) [reply packages/core/src/components/tag-input/resizableInput.tsx:31]: PR: https://github.com/palantir/blueprint/pull/5984

**Comment by adidahiya** [comment]
@nerdstep thanks for pointing out the regression. Fix here: https://github.com/palantir/blueprint/pull/5985
